util/extsort/disk_sorter.go

```diff
@@ -15,37 +15,630 @@
+	for i, file := range m.orderedFiles {
+		if bytes.Compare(file.startKey, key) > 0 {
+			m.nextFileIndex = i
+			break
+		}
+		if _, ok := openedIters[i]; ok {
+			continue
+		}
+		if bytes.Compare(file.endKey, key) <= 0 {
+			continue
+		}
+		iter, err := m.openIter(file)
+		if err != nil {
+			m.err = err
+			return false
+		}
+		if iter.Seek(key) {
+			heap.Push(&m.heap, mergingIterItem{
+				Iterator: iter,
+				index:    i,
+			})
+		} else {
+			m.err = iter.Error()
+			closeErr := iter.Close()
+			m.err = goerrors.Join(m.err, closeErr)
+			if m.err != nil {
+				return false
+			}
+		}
+	}
+	return m.maybeOpenNextFiles()
+}
+
+func (m *mergingIter) First() bool {
+	return m.Seek(nil)
+}
+
+func (m *mergingIter) Next() bool {
+	m.err = nil
+	m.curKey = append(m.curKey[:0], m.heap[0].UnsafeKey()...)
+	//revive:disable:empty-block
+	for m.next() && bytes.Equal(m.heap[0].UnsafeKey(), m.curKey) {
+	}
+	//revive:enable:empty-block
+	return m.Valid()
+}
 
-	opts   *DiskSorterOptions
-	dbDir  string        // directory for the pebble database
-	tmpDir string        // directory for temporary files
-	idGen  *atomic.Int64 // id generator for sst files
+func (m *mergingIter) next() bool {
+	if len(m.heap) == 0 {
+		return false
+	}
+	if m.heap[0].Next() {
+		heap.Fix(&m.heap, 0)
+	} else {
+		m.err = m.heap[0].Error()
+		if m.err != nil {
+			return false
+		}
+		x := heap.Pop(&m.heap)
+		item := x.(mergingIterItem)
+		m.err = item.Close()
+		if m.err != nil {
+			return false
+		}
+	}
+	return m.maybeOpenNextFiles()
+}
+
+func (m *mergingIter) maybeOpenNextFiles() bool {
+	for m.nextFileIndex < len(m.orderedFiles) {
+		file := m.orderedFiles[m.nextFileIndex]
+		if len(m.heap) > 0 && bytes.Compare(m.heap[0].UnsafeKey(), file.startKey) < 0 {
+			break
+		}
+		// The next file may overlap with the min key in the heap, so we need to
+		// open it and push it into the heap.
+		iter, err := m.openIter(file)
+		if err != nil {
+			m.err = err
+			return false
+		}
+		if !iter.First() {
+			m.err = iter.Error()
+			closeErr := iter.Close()
+			m.err = goerrors.Join(m.err, closeErr)
+			if m.err != nil {
+				return false
+			}
+		}
+		heap.Push(&m.heap, mergingIterItem{
+			Iterator: iter,
+			index:    m.nextFileIndex,
+		})
+		m.nextFileIndex++
+	}
+	return len(m.heap) > 0
+}
+
+func (m *mergingIter) Last() bool {
+	m.err = nil
+	// Close all opened iterators.
+	for _, item := range m.heap {
+		if err := item.Close(); err != nil && m.err == nil {
+			m.err = err
+		}
+	}
+	m.heap = nil
+
+	// Sort files by last key in reverse order.
+	files := slices.Clone(m.orderedFiles)
+	slices.SortFunc(files, func(a, b *fileMetadata) bool {
+		return bytes.Compare(a.lastKey, b.lastKey) > 0
+	})
+
+	// Since we don't need to implement Prev() method,
+	// we can just open the file with the largest last key.
+	for _, file := range files {
+		iter, err := m.openIter(file)
+		if err != nil {
+			m.err = err
+			return false
+		}
+		if iter.Last() {
```

Comment:
in which case `iter.Last` is false?
In other words, can we directly return the maximun value of lastKey slice?

Reply:
> in which case iter.Last is false?

Empty file.

> can we directly return the maximun value of lastKey slice?

This will affect the implementation of other methods, like `Next`, `Valid`, `UnsafeKey`, `UnsafeValue`.